pkg/tests/tasks/security/certificate/external_ca_test.go

```diff
@@ -16,101 +16,116 @@ package certificate
-		time.Sleep(time.Duration(60) * time.Second)
+		t.LogStep("Enabling control plane MTLS.")
+		oc.Patch(t, meshNamespace, "smcp", smcpName, "merge", CertSMCPPath)
+		oc.WaitSMCPReady(t, meshNamespace, smcpName)
 
-		bookinfo := examples.Bookinfo{Namespace: "bookinfo"}
-		bookinfo.Install(true)
+		t.LogStep("Installing bookinfo.")
+		bookinfo := app.Bookinfo("bookinfo")
+		bookinfo.Install(t)
+		bookinfo.WaitReady(t)
 
-		productPod, err := util.GetPodName("bookinfo", "app=productpage")
-		util.Inspect(err, "Failed to get productpage pod name", "", t)
+		if env.Getenv("SAMPLEARCH", "x86") == "p" || env.Getenv("SAMPLEARCH", "x86") == "z" {
 
-		tmpDir, err := ioutil.TempDir("", "cacerts")
-		util.Inspect(err, "Failed to create temp dir", "", t)
-		defer os.RemoveAll(tmpDir)
+			t.Log("Running Power/Z environment test.")
 
-		if env.Getenv("SAMPLEARCH", "x86") == "p" || env.Getenv("SAMPLEARCH", "x86") == "z" {
-			gatewayHTTP, _ := util.ShellSilent(`kubectl get routes -n %s istio-ingressgateway -o jsonpath='{.spec.host}'`, meshNamespace)
-			productpageURL := fmt.Sprintf("http://%s/productpage", gatewayHTTP)
-			resp, _, err := util.GetHTTPResponse(productpageURL, nil)
-			util.Inspect(err, "Failed to get HTTP Response", "", t)
-			defer util.CloseResponseBody(resp)
-		} else {
-			log.Log.Info("Verify the new certificates")
-
-			// Generate the cert files
-			util.ShellMuteOutput(`oc -n bookinfo exec %s -c istio-proxy -- openssl s_client -showcerts -connect details:9080 > %s/bookinfo-proxy-cert.txt`, productPod, tmpDir)
-			_, err = util.ShellMuteOutput(`sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' %s/bookinfo-proxy-cert.txt > %s/certs.pem`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to parse 'openssl s_client' output", "", t)
-			_, err = util.ShellMuteOutput(`awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "%s/proxy-cert-" counter ".pem"}' < %s/certs.pem`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to split certs into separate files", "", t)
-
-			// Compare them with the original certs
-			log.Log.Info("Verifying the root certificate")
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s -text -noout > %s/root-cert.crt.txt`, sampleCARoot, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s/proxy-cert-3.pem -text -noout > %s/pod-root-cert.crt.txt`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			if err := util.CompareFiles(fmt.Sprintf("%s/root-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-root-cert.crt.txt", tmpDir)); err != nil {
-				t.Errorf("Root certs do not match: %v", err)
-			}
+			t.LogStep("Checking response from productpage.")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				curl.Request(t, app.BookinfoProductPageURL(t, meshNamespace), curl.NilRequestOption{}, assert.ResponseStatus(200))
+			})
```

Comment:
Any idea why P/Z test doesn't check the certificates? Is it because sed/awk/openssl are not available? If so, we should refactor the code so that it checks the certificates using go libraries instead of external binaries.

Reply:
right, that part of the code is from an instruction doc written by Jonh.
And the reason why we didn't check that in P/Z is `openssl` was not available in their base image at that time.
@jwendell  may remember more  context.